spring-web/src/main/java/org/springframework/http/server/reactive/ReactorNetty2ServerHttpRequest.java

```diff
@@ -75,48 +76,23 @@ public ReactorNetty2ServerHttpRequest(HttpServerRequest request, Netty5DataBuffe
 
 	private static URI initUri(HttpServerRequest request) throws URISyntaxException {
 		Assert.notNull(request, "HttpServerRequest must not be null");
-		return new URI(resolveBaseUrl(request) + resolveRequestUri(request));
+		return new URI(request.scheme() + "://" + resolveHost(request) + resolveRequestUri(request));
 	}
 
-	private static URI resolveBaseUrl(HttpServerRequest request) throws URISyntaxException {
-		String scheme = getScheme(request);
-
+	private static CharSequence resolveHost(HttpServerRequest request) {
 		InetSocketAddress hostAddress = request.hostAddress();
 		if (hostAddress != null) {
-			return new URI(scheme, null, hostAddress.getHostString(), hostAddress.getPort(), null, null, null);
+			return NetUtil.toSocketAddressString(hostAddress);
```

Comment:
Before this change, the URI constructor that accepts individual URI components would validate the host and port. After this change, the value returned from `resolveHost` is not validated, the URI constructor with a String (called on line 79) doesn't either.

When I tested whether this change would help with https://github.com/spring-projects/spring-boot/issues/34395 where X-Forwarded-HOST is "localhost:3000" and X-Forwarded-Port is "3000", I saw that `resolveHost` returns "localhost:3000:3000", and the resulting URI was "http://localhost:3000:3000/hello". When I test without this change, there is an exception from the URI constructor that was previously used here because "localhost:3000" is not a valid host.